.. _custom_database_primers:

Specifying custom primers
=========================

Running prepare-reads step
--------------------------

We first ran the pipeline command with :ref:`default settings
<custom_database_defaults>`, if you skipped that we can do just the reads now:

.. code:: console

    $ mkdir -p intermediate_defaults/
    $ thapbi_pict prepare-reads -i raw_data/ -o intermediate_defaults/
    ...
    $ ls -1 intermediate_defaults/ITS1/SRR*.fasta | wc -l
    384

We then created a database from the Redekar *et al.* (2019) reference
accessions with their primers. Now we can run the pipeline again with this,
which will start by applying the prepare-reads step to the FASTQ files in
``raw_data/``:

.. code:: console

    $ mkdir -p intermediate/
    $ thapbi_pict prepare-reads -i raw_data/ -o intermediate_long/ \
      --db Redekar_et_al_2019_sup_table_3.sqlite
    ...
    $ ls -1 intermediate_long/ITS1-long/SRR*.fasta | wc -l
    384

Here the database says the left primer is ``GAAGGTGAAGTCGTAACAAGG`` (same as
the THAPBI PICT default) plus ``TTTCCGTAGGTGAACCTGCGGAAGGATCATTA`` (conserved
32bp region), and that the right primer is ``AGCGTTCTTCATCGATGTGC``. This has
reverse complement ``GCACATCGATGAAGAACGCT`` and is found about 60bp downstream
of the default right primer in *Phytophthora*, and should also match *Pythium*
and *Phytopythium* species.

i.e. We should now find the *Phytophthora* FASTA sequences extracted are about
60 - 32 = 28bp longer, and many more non-*Phytophthora* are accepted.

Will now pick a couple of samples to compare and contrast with the first run.
For clarity these examples are deliberately from the less diverse samples.
The FASTA sequences have been line wrapped at 80bp for display.

Longer sequences
----------------

We will start with ``SRR6303586`` aka ``OSU483``, a leaf-baiting sample from
a reservoir. Here is with the default primer trimming:

.. code:: console

    $ cat intermediate_defaults/ITS1/SRR6303586.fasta
    #marker:ITS1
    #left_primer:GAAGGTGAAGTCGTAACAAGG
    #right_primer:GCARRGACTTTCGTCCCYRC
    #threshold_pool:raw_data
    #raw_fastq:70396
    #flash:67870
    #cutadapt:66985
    #abundance:46117
    #threshold:100
    #max_non-spike:35109
    #max_spike-in:0
    #singletons:6207
    >11b74237bf44899f24ace62be657172a_35109
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATT
    TGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCTGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAG
    CGTTTGGGCTTCGGCTCGAACTAGTAGCTATCAATTTTAAACCCTTTCTTTAAATACTGAACATACT
    >3cd14145747f26f7461cbba643388ff6_10271
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATT
    TGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCTGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAG
    CGTTTGGGCTTCGGCTCGAACTAGTAGCTATCAATTTTAAACTCTTTCTTTAAATACTGAACATACT
    >9ba3657d5ecb7d8f35ce7f5feacbc3f7_580
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTT
    GGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCTGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGC
    GTTTGGGCTTCGGCTCGAACTAGTAGCTATCAATTTTAAACCCTTTCTTTAAATACTGAACATACT
    >5bddf030b013e783a218734230c6f542_157
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTT
    GGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCTGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGC
    GTTTGGGCTTCGGCTCGAACTAGTAGCTATCAATTTTAAACTCTTTCTTTAAATACTGAACATACT

Four very similar sequences (differing in the length of the poly-A run, seven
is more common than six, and a ``C/T`` SNP towards the end), all matched to
*Phytophthora chlamydospora* with THAPBI PICT's default settings.

With the new primer setting, which you can see listed at the start of the
header, we again get four sequences passing the abundance threshold:

.. code:: console

    $ cat intermediate_long/ITS1-long/SRR6303586.fasta
    #marker:ITS1-long
    #left_primer:GAAGGTGAAGTCGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTA
    #right_primer:AGCGTTCTTCATCGATGTGC
    #threshold_pool:raw_data
    #raw_fastq:70396
    #flash:67870
    #cutadapt:67173
    #abundance:43868
    #threshold:100
    #singletons:7468
    >e804f4fa9e197115c1f72b943e443dc7_33451
    CCACACCTAAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTTGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCC
    TGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGCGTTTGGGCTTCGGCTCGAACTAGTAGCTATC
    AATTTTAAACCCTTTCTTTAAATACTGAACATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGC
    AGTGGATGTCTAGGCTC
    >3804bc12d180cc145776cc3e77d50561_9729
    CCACACCTAAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTTGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCC
    TGTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGCGTTTGGGCTTCGGCTCGAACTAGTAGCTATC
    AATTTTAAACTCTTTCTTTAAATACTGAACATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGC
    AGTGGATGTCTAGGCTC
    >0c2bc6a79b05e53d63636509e9ea8aba_545
    CCACACCTAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTTGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCT
    GTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGCGTTTGGGCTTCGGCTCGAACTAGTAGCTATCA
    ATTTTAAACCCTTTCTTTAAATACTGAACATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGCA
    GTGGATGTCTAGGCTC
    >a09c91f2a4813209b3d22847e0b18482_143
    CCACACCTAAAAAACTTTCCACGTGAACCGTATCAACCCCTTAAATTTGGGGGCTTGCTCGGCGGCGTGCGTGCTGGCCT
    GTAATGGGTCGGCGTGCTGCTGCTGGGCAGGCTCTATCATGGGCGAGCGTTTGGGCTTCGGCTCGAACTAGTAGCTATCA
    ATTTTAAACTCTTTCTTTAAATACTGAACATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGCA
    GTGGATGTCTAGGCTC

Again four very similar sequences, each as before but with the starting
``TTTCCGTAGGTGAACCTGCGGAAGGATCATTA`` removed, and instead extended by
``GTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGCAGTGGATGTCTAGGCTC``.

The abundances are similar but slightly lower - there would have been some
minor variation in trimmed regions which would have been pooled, so with
less trimming we tend to get lower counts.

You can verify by NCBI BLAST online that the first and third (the
``C`` form) give perfect full length matches to published *Phytophthora
chlamydospora*, while an exact match to the ``T`` forms has not been
published at the time of writing (yet this occurs at good abundance in many of
these samples).

Losing sequences
----------------

If you examine ``SRR6303588`` you will see a similar example,
starting with five unique sequences (with one only just above the
default abundance threshold), dropping to four unique sequences.

Finding *Pythium*
-----------------

Now for a more interesting example, ``SRR6303596`` aka ``OSU121``, another
leaf baiting sample but from runoff water. With the defaults (using ``grep``
to omit the header):

.. code:: console

    $ grep -v "^#" intermediate_defaults/ITS1/SRR6303596.fasta
    >3dd3b5989ee07ed2d2b3fac826dbb94f_953
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAATCTTTCCACGTGAATTGTTTTGCTGTACCTTTGGG
    CTTCGCCGTTGTCTTGTTCTTTTGTAAGAGAAAGGGGGAGGCGCGGTTGGAGGCCATCAGGGGTGTGTTCGTCGCGGTTT
    GTTTCTTTTGTTGGAACTTGCGCGCGGATGCGTCCTTTTGTCAACCCATTTTTTGAATGAAAAACTGATCATACT

There was a single sequence, with no matches (NCBI BLAST suggests this is
*Phytopythium litorale*). Now with the revised primer settings this sequence
is still present but only the second most abundant sequence:

.. code:: console

    $ grep -v "^#" intermediate_long/ITS1-long/SRR6303596.fasta
    >23710597e30e5d95f1d94d6fe8848fb7_40503
    CCACACCAAAAAAACTTTCCACGTGAACCGTTGTAACTATGTTCTGTGCTCTCTTCTCGGAGAGAGCTGAACGAAGGTGG
    GCTGCTTAATTGTAGTCTGCCGATGTACTTTTAAACCCATTAAACTAATACTGAACTATACTCCGAAAACGAAAGTCTTT
    GGTTTTAATCAATAACAACTTTCAGCAGTGGATGTCTAGGCTC
    >b87b957d70d3681d0682103b0052c16b_878
    CCACACCTAAAAATCTTTCCACGTGAATTGTTTTGCTGTACCTTTGGGCTTCGCCGTTGTCTTGTTCTTTTGTAAGAGAA
    AGGGGGAGGCGCGGTTGGAGGCCATCAGGGGTGTGTTCGTCGCGGTTTGTTTCTTTTGTTGGAACTTGCGCGCGGATGCG
    TCCTTTTGTCAACCCATTTTTTGAATGAAAAACTGATCATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAA
    CTTTCAGCAGTGGATGTCTAGGCTC
    >4007e1e8dedb33b5a3c5bc2cfe67c038_388
    CCACACCAAAAAACTTTCCACGTGAACCGTTGTAACTATGTTCTGTGCTCTCTTCTCGGAGAGAGCTGAACGAAGGTGGG
    CTGCTTAATTGTAGTCTGCCGATGTACTTTTAAACCCATTAAACTAATACTGAACTATACTCCGAAAACGAAAGTCTTTG
    GTTTTAATCAATAACAACTTTCAGCAGTGGATGTCTAGGCTC
    >f2a354f8c74387a287be1d08f31df143_190
    CCACACCAAAAAAACTTTCCACGTGAACCGTTGTAACTATGTTCTGTGCTCTCTTCTCGGAGAGAGCTGAACGAAGGTGG
    GCTGCTTAATTGTAGTCTGCCGATGTACTTTTAAACCCATTAAACTAATACTGAACTATACTCCGGAAACGAAAGTCTTT
    GGTTTTAATCAATAACAACTTTCAGCAGTGGATGTCTAGGCTC
    >818d3263599c4929cf2ef4a33f952949_128
    CCACACCAAAAAAACTTTCCACGTGAACCGTTGTAACTATGTTCTGTGCTCTCTTCTCGGAGAGAGCTGAACGAAGGTGG
    GCTGCTTAATTGTAGTCTGCCGATGTACTTTTAAACCCATTAAACTAATACTGAACTATACTCCGAAAACGAAAGTCTTT
    GGTTTTAATCAATAACAACTTTCAGCAGTGGATGTCTAGGCGC
    >48bcfd8e8daaa8351cb24b7deb63a4bc_102
    CCACACCAAAAAAACTTTCCACGTGAACCGTTGTAACTATGTTCTGTGCTCTCTTCTCGGAGAGAGCTGAACGAAGGTGG
    GCTGCTTAATTGTAGTCTGCCGATGTACTTTTAAACCCATTAAACTAATACTGAACTATACTCCGAAAACGAAAGTCTTT
    GGTTTTAATCAATAACAACTTTCAGCAGTGGATGTCTAGGCCC

The probable *Phytopythium litorale* has been joined by five shorter
and very similar sequences (differing by a handful of SNPs and a
poly-A length change), which NCBI BLAST matches suggest are all
*Pythium coloratum/dissotocum*.

Finding more
------------

Another interesting example, ``SRR6303948`` aka ``OSU536.s203``,
from a runoff filtration sample. First with the default settings,
a single unique sequence matching *Phytophthora ramorum*:

.. code:: console

    $ grep -v "^#" intermediate_defaults/ITS1/SRR6303948.fasta
    >dcd6316eb77be50ee344fbeca6e005c7_1439
    TTTCCGTAGGTGAACCTGCGGAAGGATCATTACCACACCTAAAAAACTTTCCACGTGAACCGTATCAAAACCCTTAGTTG
    GGGGCTTCTGTTCGGCTGGCTTCGGCTGGCTGGGCGGCGGCTCTATCATGGCGAGCGCTTGAGCCTTCGGGTCTGAGCTA
    GTAGCCCACTTTTTAAACCCATTCCTAAATACTGAATATACT

Now with the revised primer settings, we get a further nine sequences - and
the extended *Phytophthora ramorum* sequence drops to third most abundant:

.. code:: console

    $ grep -v "^#" intermediate_long/ITS1-long/SRR6303948.fasta
    >f2d4b17eb421d8c52320c2bd883e77eb_5329
    CCACACCAAAAAAACACCCCACGTGAATTGTACTGTATGAGCTATGTGCTGCGGATTTCTGCGGCTTAGCGAAGGTTTCG
    AAAGAGACCGATGTACTTTTAAACCCCTTTACATTACTGTCTGATAAATTACATTGCAAACATTTAAAGTGGTTGCTCTT
    AATTTAACATACAACTTTCAACAGTGGATGTCTAGGCTC
    >3f03e981beda20bed3d458528fdd757e_3287
    CCACACCCGGGATCCTCGATCTTTCTCCTAGGTTAATTGTTGGGCCCTTTGAGGGTGGGCCTTAGGTGCGCTCAAGGATT
    TTTTCCTGTCCCATGTAGCTTTACTTATTTTTTTGCCTGGGTAAATGATGGATTATTTTTACAACTTTCAGCAATGGATG
    TCTAGGCTC
    >283ae6bd5fb4ba9ec5fba94a3f67b43d_1319
    CCACACCTAAAAAACTTTCCACGTGAACCGTATCAAAACCCTTAGTTGGGGGCTTCTGTTCGGCTGGCTTCGGCTGGCTG
    GGCGGCGGCTCTATCATGGCGAGCGCTTGAGCCTTCGGGTCTGAGCTAGTAGCCCACTTTTTAAACCCATTCCTAAATAC
    TGAATATACTGTGGGGACGAAAGTCTCTGCTTTTAACTAGATAGCAACTTTCAGCAGTGGATGTCTAGGCTC
    >9242fdd7b824838e583731161706caf1_438
    CCACACCAAAAAAACTTACCACGTGAATCTGTACTGTTTAGTTTTGTGCTGCGTTCGAAAGGATGCGGCTAAACGAAGGT
    TGGCTTGATTACTTCGGTAATTAGGCTGGCTGATGTACTCTTTTAAACCCCTTCATACCAAAATACTGATTTATACTGTG
    AGAATGAAAATTCTTGCTTTTAACTAGATAACAACTTTCAACAGTGGATGTCTAGGCTC
    >abf29c8b8843dcaa891031c3ba6240c5_231
    CCACACCCGGGATCCTCGATCTTTCTCCTAGGTTAATTGTTGGGCCCTTTGAGGGTGGGCCTTAGGTGCGCTCAAGGATT
    TTTTCCTGTCCCATGTAGCTTTACTTATTTTTTTGCCTGGGTAAATGATGGATTATTTTTACAACTTTCAGCAACGGATG
    TCTAGGCTC
    >1164ff4787df9ce3e411a3db5e48ebe2_230
    AATCTATCACAATCCACACCTGTGAACTTGCTTGTTGGCCTCTGCATGTGCTTCGGTATGTGCAGGTTGAGCCGATCGGA
    TTAACTTCTGGTCGGCTTGGGGCCTCAACCCAATCCTCGGATTGGTTTGGGGTCGGTCTCTATTAACAACCAACACCAAA
    CCAAACTATAAAAAAACTGAGAATGGCTTAGAGCCAAACTCACTAACCAAGACAACTCTGAACAACGGATATCTTGGCTA
    >b927746ffca4f55b1bcb0097f8ba1f4e_224
    CCACACCCGGGATCCTCGATCTTTCTCCTAGGTTAATTGTTTGGCCCTTTGAGGGTGGGCCTTAGGTGCGCTCAAGGATT
    TTTTCCTGTCCCATGTAGCTTTACTTATTTTTTTGCCTGGGTAAATGATGGATTATTTTTACAACTTTCAGCAATGGATG
    TCTAGGCTC
    >31bac939435fe6972e3e2d004937c876_189
    CCACACCTAAAAACTTTCCACGTGAATCGTTCTATATAGCTTTGTGCTTTGCGGAAACGCGAGGCTAAGCGAAGGATTAG
    CAAAGTAGTACTTCGGTGCGAAACACTTTTCCGATGTATTTTTCAAACCCTTTTACTTATACTGAACTATACTCTAAGAC
    GAAAGTCTTGGTTTTAATCCACAACAACTTTCAGCAGTGGATGTCTAGGCTC
    >7874a928f5bc50ddee39326e10ccb019_144
    CCACACCCGGGATCCTCGATCTTTCTCCTAGGTTAATTATTGGGCCCTTTGAGGGTGGGCCTTAGGTGCGCTCAAGGATT
    TTTTCCTGTCCCATGTAGCTTTACTTATTTTTTTGCCTGGGTAAATGATGGATTATTTTTACAACTTTCAGCAATGGATG
    TCTAGGCTC
    >44b31ed4182973c57683a561485745c4_102
    CCACACCAAAAAACACCCCACGTGAATTGTACTGTATGAGCTATGTGCTGCGGATTTCTGCGGCTTAGCGAAGGTTTCGA
    AAGAGACCGATGTACTTTTAAACCCCTTTACATTACTGTCTGATAAATTACATTGCAAACATTTAAAGTGGTTGCTCTTA
    ATTTAACATACAACTTTCAACAGTGGATGTCTAGGCTC

NCBI BLAST suggests some of the new sequences could be *Oomycetes*, but there
are no very close matches - and some of the tenuous best matches include
uncultured fungus, diatoms, green algae, and even green plants.
